File: scripts/github_zip_download_extract.py
#!/usr/bin/env python3
"""Download and extract a GitHub branch ZIP with proxy-aware retries."""

from __future__ import annotations

import shutil
import sys
import zipfile
import os
import socket
import time
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.request import Request, build_opener


def die(message: str) -> None:
    print(f"[github-zip-download-extract] erro: {message}", file=sys.stderr)
    raise SystemExit(1)


RETRYABLE_HTTP_STATUS = {429, 500, 502, 503, 504}
MAX_ATTEMPTS = 5
BACKOFF_SECONDS = 1.0


def download_zip(url: str, zip_path: Path, token: str) -> None:
    headers = {"User-Agent": "nvim-zip-bootstrap/1.0"}
    if token:
        headers["Authorization"] = f"Bearer {token}"

    opener = build_opener()
    last_error: Exception | None = None
    for attempt in range(1, MAX_ATTEMPTS + 1):
        request = Request(url=url, headers=headers, method="GET")
        try:
            with opener.open(request, timeout=180) as response:
                status = getattr(response, "status", 200)
                if status < 200 or status >= 300:
                    raise HTTPError(url, status, f"status {status}", response.headers, None)
                with zip_path.open("wb") as handle:
                    shutil.copyfileobj(response, handle, length=1024 * 64)
                return
        except HTTPError as error:
            last_error = error
            if error.code not in RETRYABLE_HTTP_STATUS or attempt == MAX_ATTEMPTS:
                break
        except (URLError, TimeoutError, socket.timeout, ConnectionError) as error:
            last_error = error
            if attempt == MAX_ATTEMPTS:
                break
        time.sleep(BACKOFF_SECONDS * attempt)

    die(
        "falha no download por ZIP: "
        f"{last_error}. Verifique HTTPS_PROXY/HTTP_PROXY/NO_PROXY, certificados e o host em --github-base."
    )


def extract_and_move(zip_path: Path, extract_root: Path, destination: Path, url: str) -> None:
    with zipfile.ZipFile(zip_path, "r") as zipped:
        zipped.extractall(extract_root)

    entries = [path for path in extract_root.iterdir() if not path.name.startswith(".__")]
    directories = [entry for entry in entries if entry.is_dir()]
    if len(directories) != 1:
        die(f"arquivo zip com formato inesperado: {url}")

    source = directories[0]
    destination.parent.mkdir(parents=True, exist_ok=True)
    if destination.exists() or destination.is_symlink():
        if destination.is_dir() and not destination.is_symlink():
            shutil.rmtree(destination)
        else:
            destination.unlink()
    shutil.move(str(source), str(destination))


def main() -> None:
    if len(sys.argv) != 5:
        die(
            "uso: python scripts/github_zip_download_extract.py "
            "<url> <zipPath> <extractRoot> <destination>"
        )

    url = sys.argv[1]
    zip_path = Path(sys.argv[2])
    extract_root = Path(sys.argv[3])
    destination = Path(sys.argv[4])

    zip_path.parent.mkdir(parents=True, exist_ok=True)
    extract_root.mkdir(parents=True, exist_ok=True)

    token = os.environ.get("GITHUB_TOKEN", "").strip()

    download_zip(url, zip_path, token)
    extract_and_move(zip_path, extract_root, destination, url)


if __name__ == "__main__":
    main()

File: scripts/github_zip_download_extract.mjs
#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";
import https from "node:https";
import { spawnSync } from "node:child_process";

function die(message) {
  console.error(`[github-zip-download-extract] erro: ${message}`);
  process.exit(1);
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function toErrorMessage(error) {
  if (!(error instanceof Error)) {
    return String(error);
  }
  const cause = error.cause;
  const causeMessage =
    cause && typeof cause === "object"
      ? [cause.code, cause.errno, cause.syscall, cause.hostname, cause.message]
          .filter(Boolean)
          .join(" | ")
      : "";
  return causeMessage ? `${error.message} (${causeMessage})` : error.message;
}

function buildCandidateUrls(primaryUrl) {
  return [primaryUrl];
}

function buildHttpsAgentFromProxyEnv() {
  try {
    return new https.Agent({ proxyEnv: process.env });
  } catch {
    return undefined;
  }
}

function downloadUrlToFile(url, outputPath, headers, redirectDepth = 0) {
  return new Promise((resolve, reject) => {
    if (redirectDepth > 5) {
      reject(new Error(`redirecionamento excessivo para ${url}`));
      return;
    }

    const request = https.request(
      url,
      {
        method: "GET",
        headers,
        timeout: 120000,
        agent: buildHttpsAgentFromProxyEnv(),
      },
      (response) => {
        const statusCode = response.statusCode || 0;
        const location = response.headers.location;
        if (statusCode >= 300 && statusCode < 400 && location) {
          response.resume();
          const redirectedUrl = new URL(location, url).toString();
          downloadUrlToFile(redirectedUrl, outputPath, headers, redirectDepth + 1)
            .then(resolve)
            .catch(reject);
          return;
        }
        if (statusCode < 200 || statusCode >= 300) {
          response.resume();
          reject(new Error(`falha no download (${statusCode}) para ${url}`));
          return;
        }

        const fileStream = fs.createWriteStream(outputPath);
        response.pipe(fileStream);
        fileStream.on("finish", () => {
          fileStream.close(() => resolve());
        });
        fileStream.on("error", (error) => reject(error));
      },
    );

    request.on("timeout", () => {
      request.destroy(new Error(`timeout ao baixar ${url}`));
    });
    request.on("error", (error) => reject(error));
    request.end();
  });
}

async function downloadWithRetry(urls, zipPath, headers) {
  let lastError = "";
  for (const url of urls) {
    for (let attempt = 1; attempt <= 4; attempt += 1) {
      try {
        await downloadUrlToFile(url, zipPath, headers);
        return;
      } catch (error) {
        lastError = `${url} [tentativa ${attempt}/4]: ${toErrorMessage(error)}`;
        fs.rmSync(zipPath, { force: true });
        if (attempt < 4) {
          await sleep(attempt * 1000);
        }
      }
    }
  }
  throw new Error(lastError || "falha desconhecida de download");
}

function extractZip(zipPath, extractRoot) {
  const unzipResult = spawnSync("unzip", ["-q", zipPath, "-d", extractRoot], {
    stdio: "pipe",
    encoding: "utf8",
  });
  if (unzipResult.status !== 0) {
    const reason = (unzipResult.stderr || unzipResult.stdout || "").trim();
    die(`nao foi possivel extrair zip com unzip: ${reason || "erro desconhecido"}`);
  }
}

async function main() {
  const [url, zipPath, extractRoot, destination] = process.argv.slice(2);
  if (!url || !zipPath || !extractRoot || !destination) {
    die("uso: node scripts/github_zip_download_extract.mjs <url> <zipPath> <extractRoot> <destination>");
  }

  fs.mkdirSync(path.dirname(zipPath), { recursive: true });
  fs.mkdirSync(extractRoot, { recursive: true });

  const headers = { "User-Agent": "nvim-zip-bootstrap/1.0" };
  const token = (process.env.GITHUB_TOKEN || "").trim();
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }

  const candidateUrls = buildCandidateUrls(url);
  try {
    await downloadWithRetry(candidateUrls, zipPath, headers);
    extractZip(zipPath, extractRoot);

    const entries = fs
      .readdirSync(extractRoot, { withFileTypes: true })
      .filter((entry) => !entry.name.startsWith(".__"))
      .map((entry) => path.join(extractRoot, entry.name));
    const dirs = entries.filter((entryPath) => fs.statSync(entryPath).isDirectory());
    if (dirs.length !== 1) {
      die(`arquivo zip com formato inesperado: ${url}`);
    }

    const source = dirs[0];
    fs.mkdirSync(path.dirname(destination), { recursive: true });
    fs.rmSync(destination, { recursive: true, force: true });
    fs.renameSync(source, destination);
    return;
  } catch (downloadError) {
    const message = toErrorMessage(downloadError);
    die(
      [
        `falha no download por ZIP: ${message}`,
        "fallback por git clone foi desabilitado para ambientes corporativos com restricao de organizacao.",
        "configure um espelho interno e use --github-base <url> (ou GITHUB_BASE_URL) para apontar para esse host.",
      ].join(" "),
    );
  }
}

main().catch((error) => {
  const message = error instanceof Error ? error.message : String(error);
  die(message || "erro desconhecido");
});
